app/services/onestop_id.rb

```diff
@@ -1,51 +1,119 @@
+
   COMPONENT_SEPARATOR = '-'
   GEOHASH_FILTER = /[^0123456789bcdefghjkmnpqrstuvwxyz]/
   NAME_TILDE = /[\-\:\&\@\/]/
   NAME_FILTER = /[^a-zA-Z\d\@\~]/
   IDENTIFIER_TEMPLATE = Addressable::Template.new("gtfs://{feed_onestop_id}/{entity_prefix}/{entity_id}")
 
-  attr_accessor :entity_prefix, :geohash, :name
+  class OnestopIdBase
 
-  def initialize(string: nil, entity_prefix: nil, geohash: nil, name: nil)
-    if string && string.length > 0
-      @entity_prefix = string.split(COMPONENT_SEPARATOR)[0]
-      @geohash = string.split(COMPONENT_SEPARATOR)[1]
-      @name = string.split(COMPONENT_SEPARATOR)[2]
-    elsif entity_prefix && geohash && name
-      # Filter geohash and name; validate later
-      @entity_prefix = entity_prefix
-      @geohash = geohash_filter(geohash)
-      @name = name_filter(name)
-    else
-      raise ArgumentError.new('either a string or entity/geohash/name must be specified')
+    PREFIX = nil
+    MODEL = nil
+    NUM_COMPONENTS = 3
+
+    attr_accessor :geohash, :name
+
+    def initialize(string: nil, geohash: nil, name: nil, entity_prefix: nil)
+      if string && string.length > 0
+        @geohash = string.split(COMPONENT_SEPARATOR)[1]
+        @name = string.split(COMPONENT_SEPARATOR)[2]
+      elsif geohash && name
+        # Filter geohash and name; validate later
+        @geohash = geohash_filter(geohash)
+        @name = name_filter(name)
+      else
+        raise ArgumentError.new('either a string or entity/geohash/name must be specified')
+      end
+      # Check valid OnestopID
+      is_a_valid_onestop_id, errors = self.validate(self.to_s)
+      if !is_a_valid_onestop_id
+        raise ArgumentError.new(errors.join(', '))
+      end
+      self
+    end
+
+    def to_s
+      [self.class::PREFIX, @geohash, @name].join(COMPONENT_SEPARATOR)
+    end
+
+    def validate(value)
+      split = value.split(COMPONENT_SEPARATOR)
+      errors = []
+      errors << 'must not be empty' if value.blank?
+      errors << 'incorrect length' unless split.size == self.class::NUM_COMPONENTS
+      errors << 'invalid geohash' unless validate_geohash(split[1])
+      errors << 'invalid name' unless validate_name(split[2])
+      # errors << 'invalid suffix' unless validate_suffix(split[3])
+      return (errors.size == 0), errors
+    end
+
+    def validate_geohash(value)
+      !(value =~ GEOHASH_FILTER)
+    end
+
+    def validate_name(value)
+      !(value =~ NAME_FILTER)
+    end
+
+    def validate_prefix(value)
+      value == self.PREFIX
+    end
+
+    def validate_suffix(value)
+      true
+    end
+
+    private
+
+    def name_filter(value)
+      value.downcase.gsub(NAME_TILDE, '~').gsub(NAME_FILTER, '')
     end
-    # Check valid OnestopID
-    is_a_valid_onestop_id, errors = OnestopId.validate_onestop_id_string(self.to_s)
-    if !is_a_valid_onestop_id
-      raise ArgumentError.new(errors.join(', '))
+
+    def geohash_filter(value)
+      value.downcase.gsub(GEOHASH_FILTER, '')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.